javascriptv3/example_code/s3/src/s3_multipartupload.js

```diff
@@ -1,95 +1,84 @@
+} from "@aws-sdk/client-s3";
+import { s3Client } from "./libs/s3Client.js";
 
-// Set the parameters.
-export const createParams = {
-  Bucket: "BUCKET",
-  Key: "LARGE_FILE_NAME",
-};
+const twentyFiveMB = 25 * 1024 * 1024;
 
-// Specify how many parts in the upload. Between 1 and 10000.
-const parts = NUMBER_OF_PARTS; // For example, 3.
+export const createString = (size = twentyFiveMB) => {
+  return "x".repeat(size);
+};
 
 export const run = async () => {
+  const bucketName = process.env.S3_BUCKET_NAME; // S3_BUCKET_NAME="BUCKET_NAME" && node s3_multipartupload.js
+  const key = "multipart.txt";
+  const str = createString();
+  const buffer = Buffer.from(str, "utf8");
+
   try {
-    // Create the mutlipart upload.
-    const data = await s3Client.send(
-      new CreateMultipartUploadCommand(createParams)
+    const multipartUpload = await s3Client.send(
+      new CreateMultipartUploadCommand({
+        Bucket: bucketName,
+        Key: key,
+      })
     );
-    console.log("Upload started. Upload ID: ", data.UploadId);
-    return data
-    // Use loop to run UploadPartCommand for each part.
-    var i;
-    for (i = 0; i < parts; i++) {
-      var uploadParams = {
-        Bucket: createParams.Bucket,
-        Key: createParams.Key,
-        PartNumber: i,
-        UploadId: data.UploadId,
-      };
-      try {
-        const data = await s3Client.send(new UploadPartCommand(uploadParams));
-        console.log("Part uploaded. ETag: ", data.ETag);
-        return data; // For unit tests.
-        var completeParams = {
-          Bucket: createParams.Bucket,
-          Key: createParams.Key,
-          MultipartUpload: {
-            Parts: [
-              {
-                ETag: data.ETag,
-                PartNumber: i,
-              },
-            ],
-          },
-          UploadId: uploadParams.UploadId,
-        };
-      } catch (err) {
-        console.log("Error uploading part", err);
-      }
+
+    const uploadPromises = [];
+    // Multipart uploads require a minimum size of 5 MB per part.
+    const partSize = Math.ceil(buffer.length / 5);
+
+    // Upload each part.
+    for (let i = 0; i < 5; i++) {
+      const start = i * partSize;
+      const end = start + partSize;
+      uploadPromises.push(
+        s3Client
+          .send(
+            new UploadPartCommand({
+              Bucket: bucketName,
+              Key: key,
+              UploadId: multipartUpload.UploadId,
+              Body: buffer.subarray(start, end),
+              PartNumber: i + 1,
+            })
+          )
+          .then((d) => {
+            console.log("Part", i + 1, "uploaded");
+            return d;
+          })
+      );
     }
-  } catch (err) {
-    console.log("Error creating upload ", err);
-  }
-  try {
-    // Complete the mutlipart upload.
-    const data = await s3Client.send(
-      new CompleteMultipartUploadCommand(completeParams)
+
+    const uploadResults = await Promise.all(uploadPromises);
```

Comment:
You complete the upload even if a part fails. You should probably catch that, roll back the upload, and tell the user.

Edit: I see the catch, now. Let's normalize doing something useful & explicit on failure?

Reply:
Added a step to try and abort the multipart upload if it was started.